Unsuitable Method for URL (Spring Boot 3.3.2 and Flyway)

I'm using the following dependencies for a multi-module maven project

datasource-proxy-spring-boot-starter - 1.9.1
Java 21
Spring Boot 3.3.2 (also tried with 3.3.1)
Postgres DB
Flyway

When I start the application, I receive the following exception. I don't have the problem when not using datasource-proxy-spring-boot-starter. 

Caused by: org.springframework.boot.jdbc.UnsupportedDataSourcePropertyException: Unable to find suitable method for url
	at org.springframework.boot.jdbc.UnsupportedDataSourcePropertyException.throwIf(UnsupportedDataSourcePropertyException.java:36) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.jdbc.DataSourceBuilder$ReflectionDataSourceProperties.getMethod(DataSourceBuilder.java:580) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.jdbc.DataSourceBuilder$ReflectionDataSourceProperties.get(DataSourceBuilder.java:570) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.jdbc.DataSourceBuilder.build(DataSourceBuilder.java:180) ~[spring-boot-3.3.2.jar:3.3.2]
	at org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration$FlywayConfiguration.getMigrationDataSource(FlywayAutoConfiguration.java:205) ~[spring-boot-autoconfigure-3.3.2.jar:3.3.2]
	at org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration$FlywayConfiguration.configureDataSource(FlywayAutoConfiguration.java:184) ~[spring-boot-autoconfigure-3.3.2.jar:3.3.2]
	at org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration$FlywayConfiguration.flyway(FlywayAutoConfiguration.java:172) ~[spring-boot-autoconfigure-3.3.2.jar:3.3.2]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:146) ~[spring-beans-6.1.11.jar:6.1.11]

--- a/datasource-decorator-spring-boot-autoconfigure/build.gradle.kts
+++ b/datasource-decorator-spring-boot-autoconfigure/build.gradle.kts
@@ -42,6 +42,7 @@ dependencies {
     testImplementation("org.apache.commons:commons-dbcp2:2.9.0")
     testImplementation("org.apache.tomcat:tomcat-jdbc:10.1.5")
     testImplementation("com.zaxxer:HikariCP:5.0.1")
+    testImplementation("org.flywaydb:flyway-core:9.5.1")
 }
 
 tasks {

--- a/datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/boot/jdbc/decorator/DecoratedDataSource.java
+++ b/datasource-decorator-spring-boot-autoconfigure/src/main/java/com/github/gavlyukovskiy/boot/jdbc/decorator/DecoratedDataSource.java
@@ -20,6 +20,7 @@ import org.springframework.jdbc.datasource.DelegatingDataSource;
 
 import javax.sql.DataSource;
 
+import java.sql.SQLException;
 import java.util.List;
 import java.util.stream.Collectors;
 
@@ -81,6 +82,24 @@ public class DecoratedDataSource extends DelegatingDataSource {
      */
     public List<DataSourceDecorationStage> getDecoratingChain() {
         return decoratingChain;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T> T unwrap(Class<T> iface) throws SQLException {
+        // Spring Boot unwrapping simply passes 'unwrap(DataSource.class)' expecting real datasource to be returned
+        // if the real datasource type matches - return real datasource
+        if (iface.isInstance(getRealDataSource())) {
+            return (T) getRealDataSource();
+        }
+        // As some decorators don't consider their types during unwrapping
+        // if their type is specifically requested, we can return the decorator itself
+        for (DataSourceDecorationStage dataSourceDecorationStage : decoratingChain) {
+            if (iface.isInstance(dataSourceDecorationStage.getDataSource())) {
+                return (T) dataSourceDecorationStage.getDataSource();
+            }
+        }
+        return super.unwrap(iface);
     }
 
     @Override

--- a/datasource-decorator-spring-boot-autoconfigure/src/test/java/com/github/gavlyukovskiy/boot/jdbc/decorator/DataSourceDecoratorAutoConfigurationTests.java
+++ b/datasource-decorator-spring-boot-autoconfigure/src/test/java/com/github/gavlyukovskiy/boot/jdbc/decorator/DataSourceDecoratorAutoConfigurationTests.java
@@ -27,6 +27,7 @@ import org.junit.jupiter.api.Test;
 import org.springframework.beans.DirectFieldAccessor;
 import org.springframework.boot.autoconfigure.AutoConfigurations;
 import org.springframework.boot.autoconfigure.context.PropertyPlaceholderAutoConfiguration;
+import org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration;
 import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
 import org.springframework.boot.test.context.runner.ApplicationContextRunner;
 import org.springframework.context.annotation.Bean;
@@ -251,6 +252,35 @@ class DataSourceDecoratorAutoConfigurationTests {
             DataSource dataSource = context.getBean(DataSource.class);
             assertThat(dataSource).isNotInstanceOf(DecoratedDataSource.class);
             assertThat(dataSource).isInstanceOf(AbstractRoutingDataSource.class);
+        });
+    }
+
+    @Test
+    void testUnwrapsRealDataSourceForFlyway() {
+        ApplicationContextRunner contextRunner = this.contextRunner
+                .withConfiguration(AutoConfigurations.of(FlywayAutoConfiguration.class))
+                .withPropertyValues("spring.flyway.user=sa");
+
+        contextRunner.run(context -> {
+            DataSource dataSource = context.getBean(DataSource.class);
+            assertThat(dataSource).isInstanceOf(DecoratedDataSource.class);
+            assertThat(dataSource.unwrap(HikariDataSource.class)).isInstanceOf(HikariDataSource.class);
+            assertThat(dataSource.unwrap(DataSource.class)).isInstanceOf(HikariDataSource.class);
+        });
+    }
+
+    @Test
+    void testUnwrapsProxyDataSourcesFromChain() {
+        ApplicationContextRunner contextRunner = this.contextRunner
+                .withConfiguration(AutoConfigurations.of(FlywayAutoConfiguration.class))
+                .withPropertyValues("spring.flyway.user=sa");
+
+        contextRunner.run(context -> {
+            DataSource dataSource = context.getBean(DataSource.class);
+            assertThat(dataSource).isInstanceOf(DecoratedDataSource.class);
+            assertThat(dataSource.unwrap(ProxyDataSource.class)).isInstanceOf(ProxyDataSource.class);
+            assertThat(dataSource.unwrap(P6DataSource.class)).isInstanceOf(P6DataSource.class);
+            assertThat(dataSource.unwrap(FlexyPoolDataSource.class)).isInstanceOf(FlexyPoolDataSource.class);
         });
     }
 

